### PR TITLE
Exclude test data from the Python package

### LINKED ISSUE #5925: PyPI package contains test data (40MB unpacked)
### Describe the bug

The ONNX package on PyPI contains all test files found at https://github.com/onnx/onnx/tree/main/onnx/backend/test/data . These constitute ~40MB unpacked or more than 70% of the total package size. 

### System information
I checked the 1.15.0 MacOS wheel, but judging by the compressed file size all platforms are affected: https://pypi.org/project/onnx/#files

### Expected behavior
These test files should not be installed in a production environment.

### Other notes

Are there any plans to move those binary files out of git / generate them on the fly?

### PR DESCRIPTION
**Pull request by cbourjau**
### Description
`onnx/backend/test/data` contains large test files which should not be included in the PyPI package since they are irrelvant to the end user. This PR simply excludes them when building the package. The files remain available for running the test cases. This reduces the size of the built `onnx` package from 51MB to 12MB (uncompressed).

### Motivation and Context

Shipping a package that is more than 4x larger than necessary is not good. More discussion and links to previous discussions around this issue can be found at #5925. 

Closes #5925

### CODE REVIEW

**Comment by andife** [comment]
@cjvolzka I think that would be good for 1.16? At least to establish that nobody really uses it in this way? But shouldn't that actually be the case?

**Comment by cjvolzka** [comment]
@andife My initial thought is that this missed the boat for the 1.16 release.  I'm working on uploading 1.16.0rc1 later today and this hasn't even merged into `main` yet. 

I don't _think_ anyone would be using these test models but I can't say for sure. Since there's uncertainty on that, if we removed them, I'd want it in rc1. Otherwise I fear someone may validate our rc1, and if that goes well, they may not try rc2, and then miss that this somehow breaks them.

So getting into 1.16 seems like too much of a rush job for an issue that's already existed for a while. Yes we should definitely fix it, but I lean towards not rushing it just to get it in now.

I could be swayed if some of the other primary reviewers disagree and want it in, but that's my hot take.

**Comment by gramalingam** [comment]
I agree with @cjvolzka ... furthermore, this is still a draft. I agree it is a good idea to do this, but not rush this (just like Charles says).

**Comment by cbourjau** [comment]
I agree that given how long the data was packaged in the PyPI distribution one more release is not going to make a big difference.

**Comment by postrational** [comment]
Reducing the size of the ONNX package does sound like a good idea. However, there are scenarios which do utilize the package in order to run testing. In particular, the [ONNX Scoreboard](https://github.com/onnx/backend-scoreboard) relies on installing ONNX packages as they are released to run tests on submitted backends.

In addition, other users may have based their CI tests on this approach or downloading the ONNX package as a test platform. We need to provide a well defined upgrade path for these users, as this is a breaking change for them.

**Comment by cbourjau** [comment]
I cleaned up the last CI issues and marked this as "ready for review". As @postrational noted, this is indeed a breaking change, but I hardly see a way around it. I see three possible ways forward for downstream projects:
- Migrate to a git-submodule setup
- Generate the test files on the fly (much more work)
- Vendor the files via a separate package such as `onnx-test-data`

File: onnx/backend/test/loader/__init__.py [hunk 1]
```diff
@@ -8,13 +8,9 @@
...
 from onnx.backend.test.case.test_case import TestCase
 
-DATA_DIR = os.path.join(
-    os.path.dirname(os.path.realpath(os.path.dirname(__file__))), "data"
-)
-
 
 def load_model_tests(
-    data_dir: str = DATA_DIR,
+    data_dir: str,
```
Comment (justinchuby) [inline onnx/backend/test/loader/__init__.py:13]:
```suggestion
    data_dir: str | os.PathLike,
```

we can make it take a pathlike object

File: onnx/backend/test/runner/__init__.py [hunk 2]
```diff
@@ -57,6 +58,7 @@ class Runner:
     def __init__(
         self,
         backend: type[Backend],
+        test_data_dir: Path,
```
Comment (justinchuby) [inline onnx/backend/test/runner/__init__.py:61]:
```suggestion
        test_data_dir: str | os.PathLike,
```

File: tests/backend/test_backend_reference.py [hunk 3]
```diff
@@ -18,6 +18,9 @@
 from onnx import ModelProto
 from onnx.backend.base import Device, DeviceType
 from onnx.reference import ReferenceEvaluator
+from onnx.test import version_utils
```
Comment (justinchuby) [inline tests/backend/test_backend_reference.py:21]: Is onnx.test part of the package? Should we create a copy of  version_utils?

File: tests/backend/test_backend_reference.py [hunk 4]
```diff
@@ -18,6 +18,9 @@
...
+
+TEST_DATA_DIR = Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```
Comment (justinchuby) [inline tests/backend/test_backend_reference.py:23]:
```suggestion
_TEST_DATA_DIR = Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```

File: tests/backend/test_backend_test.py [hunk 5]
```diff
@@ -29,6 +30,8 @@
 # We don't enable report in this test because the report collection logic itself
 # fails when models are mal-formed.
 
+TEST_DATA_DIR = Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```
Comment (justinchuby) [inline tests/backend/test_backend_test.py:33]:
```suggestion
_TEST_DATA_DIR = Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```

File: tests/backend/test_backend_test.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 import os
 import platform
 import unittest
+from pathlib import Path
```
Comment (justinchuby) [inline tests/backend/test_backend_test.py:9]:
```suggestion
import pathlib
```

Import modules only: https://google.github.io/styleguide/pyguide.html#224-decision

File: tests/backend/test_backend_reference.py [hunk 7]
```diff
@@ -6,10 +6,10 @@
 import platform
 import sys
 import unittest
+from pathlib import Path
```
Comment (justinchuby) [inline tests/backend/test_backend_reference.py:9]:
```suggestion
import pathlib
```

Import modules only: https://google.github.io/styleguide/pyguide.html#224-decision

**Review by justinchuby — COMMENTED** [review]
Thanks! LGTM; would like to let others chime in. There may be some documentation we need to update as well.

File: tests/backend/test_backend_reference.py (hunk 3 above)
Comment (cbourjau) [reply tests/backend/test_backend_reference.py]: It was literally a one-liner so I just copied it and removed this import.

File: onnx/backend/test/runner/__init__.py [hunk 8]
```diff
@@ -66,26 +68,31 @@ def __init__(
...
+        self._test_data_dir = test_data_dir
 
         # This is the source of the truth of all test functions.
         # Properties `test_cases`, `test_suite` and `tests` will be
         # derived from it.
         # {category: {name: func}}
         self._test_items: dict[str, dict[str, TestItem]] = defaultdict(dict)
 
-        for rt in load_model_tests(kind="node"):
+        for rt in load_model_tests(data_dir=str(test_data_dir), kind="node"):
```
Comment (justinchuby) [inline onnx/backend/test/runner/__init__.py:79]:
```suggestion
        for rt in load_model_tests(data_dir=test_data_dir, kind="node"):
```

and below

File: onnx/backend/test/stat_coverage.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import os
+from pathlib import Path
```
Comment (justinchuby) [inline onnx/backend/test/stat_coverage.py:8]:
```suggestion
import pathlib
```

File: onnx/backend/test/stat_coverage.py [hunk 10]
```diff
@@ -157,7 +153,9 @@ def gen_model_test_coverage(
     # Need to grab associated nodes
     attrs: Dict[str, Dict[str, List[Any]]] = {}
     model_paths: List[Any] = []
-    for rt in load_model_tests(kind="real"):
+    for rt in load_model_tests(
+        data_dir=str(Path(__file__).parent / "data"), kind="real"
```
Comment (justinchuby) [inline onnx/backend/test/stat_coverage.py:157]:
```suggestion
        data_dir=path.Path(__file__).parent / "data", kind="real"
```

File: tests/backend/test_backend_onnxruntime.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 
 import platform
 import unittest
+from pathlib import Path
```
Comment (justinchuby) [inline tests/backend/test_backend_onnxruntime.py:9]:
```suggestion
import pathlib
```

File: tests/backend/test_backend_onnxruntime.py [hunk 12]
```diff
@@ -26,6 +27,7 @@
     ort: Any = None  # type: ignore[no-redef]
     ort_version: Any = None  # type: ignore[no-redef]
 
+TEST_DATA_DIR = Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```
Comment (cbourjau) [inline tests/backend/test_backend_onnxruntime.py:30]:
```suggestion
TEST_DATA_DIR = pathlib.Path(__file__).parent.parent.parent / "onnx/backend/test/data"
```

File: onnx/backend/test/runner/__init__.py [hunk 13]
```diff
@@ -15,6 +15,7 @@
 import time
 import unittest
 from collections import defaultdict
+from pathlib import Path
```
Comment (justinchuby) [inline onnx/backend/test/runner/__init__.py:18]: Please make sure it is cleaned up. Otherwise lgtm.

**Review by justinchuby — APPROVED** [review]

File: onnx/backend/test/runner/__init__.py (hunk 13 above)
Comment (justinchuby) [reply onnx/backend/test/runner/__init__.py:18]: ?

Comment (cbourjau) [reply onnx/backend/test/runner/__init__.py]: Apologies! I was relying on the linter CI step which passed after you made this comment!

Comment (justinchuby) [reply onnx/backend/test/runner/__init__.py]: No worries. Yeah unfortunately some aspects are not covered by the CI

**Comment by justinchuby** [comment]
Reopening to trigger test

File: pyproject.toml [hunk 14]
```diff
@@ -43,16 +43,21 @@ onnx = [
     "**/*.h",
     "**/*.proto",
     "**/*.pyi",
-    "backend/test/data/**/*",
     "py.typed",
 ]
 
+[tool.setuptools.exclude-package-data]
```
Comment (xadupre) [inline pyproject.toml:49]: I think the change goes into the right direction. My only concern is runtimes such as onnxruntime expects to find this data. I'd me more confortable to display an error message saying what instruction is needed to get the data and have the backend tests passing.

**Comment by gramalingam** [comment]
Combining the suggestions made by @cbourjau  and @postrational : providing a separate testdata package, while removing it from the main package, makes sense to me.

It comes down to a question of who is willing to do that (create the new testdata package): it would be great if we have a volunteer to do this!

There's an interesting versioning complication to this issue: due to the way tests are defined and generated, the test-cases generated in different releases of onnx (from the same test-case source) may differ (using different opset-versions of the op). Eg., onnx release N and N+1 may have versions of the same test-case, but using different opset-versions (if the particular op's version was bumped in release N+1). Ideally, we should provide a better API/interface to allow users to get the versions of these test-cases they want from a single package. See [here](https://github.com/onnx/onnx/blob/5ecc0a9801b724bd71859ceacd1562cfed4c415c/onnx/backend/test/case/node/__init__.py#L274)

File: pyproject.toml (hunk 14 above)
Comment (cbourjau) [reply pyproject.toml:49]: onnxruntime already fetches the entire onnx source during the build process anyway. Are they really using the PyPI package for testing, too?

**Comment by cbourjau** [comment]
I'm not sure it needs to be PyPI package at the end of the day. The Array API standard has a phenomenal testing [project](https://github.com/data-apis/array-api-tests) that (a) does not ship with binaries and (b) is not a PyPI package. Having something like that for ONNX would be ideal (but a lot of work), IMHO.

File: pyproject.toml (hunk 14 above)
Comment (xadupre) [reply pyproject.toml:49]: Yes, it is used. onnx should be shipped without data. That's not a debate. Since this data can be automatically generated, we could also generate this data when the package is being installed. Something like ``pip install onnx <options>``... So developers using the data could just install onnx with this option and nothing else would change.

**Comment by mgehre-amd** [comment]
We also use the test models from the `onnx` package to test our ONNX compilers against. I'd ask you to please continue shipping this either inside the onnx package or in another python wheel.

**Comment by justinchuby** [comment]
> We also use the test models from the `onnx` package to test our ONNX compilers against. I'd ask you to please continue shipping this either inside the onnx package or in another python wheel.

Would it be a solution if you cloned the onnx repository to get the test data?

**Comment by mgehre-amd** [comment]
> > We also use the test models from the `onnx` package to test our ONNX compilers against. I'd ask you to please continue shipping this either inside the onnx package or in another python wheel.
> 
> Would it be a solution if you cloned the onnx repository to get the test data?

If I could run ` onnx.backend.test` without any additional build step (just changing PYTHONPATH), that would be fine for us.

**Comment by justinchuby** [comment]
> > > We also use the test models from the `onnx` package to test our ONNX compilers against. I'd ask you to please continue shipping this either inside the onnx package or in another python wheel.
> > 
> > 
> > Would it be a solution if you cloned the onnx repository to get the test data?
> 
> If I could run ` onnx.backend.test` without any additional build step (just changing PYTHONPATH), that would be fine for us.

Could you share an example usage just to make sure we understand it clearly? (A script etc.)

**Comment by mgehre-amd** [comment]
We use the (I hope) common way of running the backend tests via
```
import sys
import os
import unittest
import onnx.backend.test

import OurBackend

# A pytest magic variable to load pytest_report.py from this directory.
pytest_plugins = ("pytest_report",)

backend_test = onnx.backend.test.runner.Runner(OurBackend, __name__)
globals().update(backend_test.enable_report().test_cases)

if __name__ == "__main__":
    unittest.main()
```

**Comment by xadupre** [comment]
I don't know what the status is for this PR. On my side, I still think it is useful. An easy way would be to have two package, onnx and onnx-data. If both are installed, nothing would change for the user. If the second one is not installed, anything based the data would fail.

**Comment by cbourjau** [comment]
@xadupre I think that is a nice way to go about it! Ideally, `onnx-data` could also be migrated into it's own repository to avoid adding yet more binary data to this repository which is already very large.

**Comment by jameslamb** [comment]
> *This reduces the size of the built onnx package from 51MB to 12MB (uncompressed).*

@cbourjau I found this PR from looking at your profile after you opened https://github.com/microsoft/LightGBM/pull/6478. In case it's helpful, I want to mention... I've been working on a CLI tool for checking Python packages that can be used to enforce in CI, [among other things](https://pydistcheck.readthedocs.io/en/latest/check-reference.html), a limit like "the compressed size of this wheel should not be larger than `n`".

It's called `pydistcheck`: https://github.com/jameslamb/pydistcheck

We use that in LightGBM and it's caught a few "we're accidentally including these unnecessary files" or "these debug symbols aren't being stripped correctly" types of issues, like this:

```shell
pydistcheck \
    --inspect \
    --max-allowed-size-compressed '5M' \
    --max-allowed-size-uncompressed '15M' \
    --max-allowed-files 800 \
    dist/*
```
 ([code link](https://github.com/microsoft/LightGBM/blob/ad1237d90830d795f450f40a8697f8bc6faa4d1a/.ci/check_python_dists.sh#L31-L45))

But to be fair... basically no one else that I'm aware of is using it 😬 

Just mentioning in case you'd find it helpful for ensuring future changes in this project don't accidentally undo the size-reduction work you've done here.

**Comment by justinchuby** [comment]
> I don't know what the status is for this PR. On my side, I still think it is useful. An easy way would be to have two package, onnx and onnx-data. If both are installed, nothing would change for the user. If the second one is not installed, anything based the data would fail.


> @xadupre I think that is a nice way to go about it! Ideally, `onnx-data` could also be migrated into it's own repository to avoid adding yet more binary data to this repository which is already very large.

I like this idea. We can discuss creating the repository. @cbourjau would you be interested in initiating the repository?

**Comment by cbourjau** [comment]
I'm on holidays right now but am happy to take another look at this in about two weeks time.

**Comment by cbourjau** [comment]
I spent some more time thinking about this issue and I'm not sure if moving `onnx/backend/test/data` into its own repository and/or package is worth the complication. There are various considerations:
- Having (parts of the tests) in a different repository significantly complicates contributions of new tests and operators
- Splitting the test data out of the regular package while still leaving the public API unchanged:
  -  A default `onnx` installation would be in an inconsistent state where some public functions throw IO errors
  - Complicates the already complicated release and packaging story
  - Is still a SemVer breaking change 
  - Even after the breaking change discussed here the test architecture is still not in great shape. A general non-backward compatible refactoring of it would remain desirable, but than we would need to move through two breaking changes.
 
 In essence I think that we should either rip of the bandage and completely remove `onnx.backend.test` from the public API of the PyPI `onnx` package (without providing a PyPI package with the test data due to the maintenance cost), or we don't rock the boat in this regard until we we come up with a generally better test architecture.

**Comment by justinchuby** [comment]
Thanks for thinking this through! Even though I don’t love the growing size of the package due to the test data, I think removing the public test apis is too disruptive to current users. 

I am in favor of a better testing architecture. Our current tests do not cover all opsets and all data types, and is limited in terms of the edge cases they provide.

**Comment by andife** [comment]
@saurabhtangri @cbourjau At least for an upcoming onnx 2.0 we should definitely discuss this clearly again

**Comment by cbourjau** [comment]
Closed in favor of #7959
